File: app/src/main/java/com/dpis/module/PackageConfigSnapshot.java
package com.dpis.module;

import com.dpis.module.viewport.ViewportApplyMode;
import com.dpis.module.viewport.ViewportTargetSpec;

import com.dpis.module.hooks.HookDomainOverride;

final class PackageConfigSnapshot {
    final String packageName;
    final boolean dpisEnabled;
    final ViewportTargetSpec targetViewportSpec;
    final String targetViewportMode;
    final Integer targetFontScalePercent;
    final String targetFontMode;
    final String targetTypefaceId;
    final boolean flutterFontHookEnabled;
    final boolean flutterSettingsFontHookEnabled;
    final boolean hyperOsFlutterFontHookEnabled;
    final HookDomainOverride hookDomainOverride;

    PackageConfigSnapshot(String packageName,
                          boolean dpisEnabled,
                          Integer targetViewportWidthDp,
                          String targetViewportMode,
                          Integer targetFontScalePercent,
                          String targetFontMode,
                          String targetTypefaceId,
                          boolean flutterFontHookEnabled,
                          boolean flutterSettingsFontHookEnabled,
                          boolean hyperOsFlutterFontHookEnabled) {
        this(packageName,
                dpisEnabled,
                targetViewportWidthDp != null
                        ? ViewportTargetSpec.absoluteDp(targetViewportWidthDp)
                        : ViewportTargetSpec.off(),
                targetViewportMode,
                targetFontScalePercent,
                targetFontMode,
                targetTypefaceId,
                flutterFontHookEnabled,
                flutterSettingsFontHookEnabled,
                hyperOsFlutterFontHookEnabled,
                HookDomainOverride.automatic());
    }

    PackageConfigSnapshot(String packageName,
                          boolean dpisEnabled,
                          Integer targetViewportWidthDp,
                          String targetViewportMode,
                          Integer targetFontScalePercent,
                          String targetFontMode,
                          String targetTypefaceId,
                          boolean flutterFontHookEnabled,
                          boolean flutterSettingsFontHookEnabled,
                          boolean hyperOsFlutterFontHookEnabled,
                          HookDomainOverride hookDomainOverride) {
        this(packageName,
                dpisEnabled,
                targetViewportWidthDp != null
                        ? ViewportTargetSpec.absoluteDp(targetViewportWidthDp)
                        : ViewportTargetSpec.off(),
                targetViewportMode,
                targetFontScalePercent,
                targetFontMode,
                targetTypefaceId,
                flutterFontHookEnabled,
                flutterSettingsFontHookEnabled,
                hyperOsFlutterFontHookEnabled,
                hookDomainOverride);
    }

    PackageConfigSnapshot(String packageName,
                          boolean dpisEnabled,
                          ViewportTargetSpec targetViewportSpec,
                          Integer targetViewportWidthDp,
                          String targetViewportMode,
                          Integer targetFontScalePercent,
                          String targetFontMode,
                          String targetTypefaceId,
                          boolean flutterFontHookEnabled,
                          boolean flutterSettingsFontHookEnabled,
                          boolean hyperOsFlutterFontHookEnabled,
                          HookDomainOverride hookDomainOverride) {
        this(packageName,
                dpisEnabled,
                targetViewportSpec != null
                        ? targetViewportSpec
                        : (targetViewportWidthDp != null
                                ? ViewportTargetSpec.absoluteDp(targetViewportWidthDp)
                                : ViewportTargetSpec.off()),
                targetViewportMode,
                targetFontScalePercent,
                targetFontMode,
                targetTypefaceId,
                flutterFontHookEnabled,
                flutterSettingsFontHookEnabled,
                hyperOsFlutterFontHookEnabled,
                hookDomainOverride);
    }

    PackageConfigSnapshot(String packageName,
                          boolean dpisEnabled,
                          ViewportTargetSpec targetViewportSpec,
                          String targetViewportMode,
                          Integer targetFontScalePercent,
                          String targetFontMode,
                          String targetTypefaceId,
                          boolean flutterFontHookEnabled,
                          boolean flutterSettingsFontHookEnabled,
                          boolean hyperOsFlutterFontHookEnabled,
                          HookDomainOverride hookDomainOverride) {
        this.packageName = packageName;
        this.dpisEnabled = dpisEnabled;
        this.targetViewportSpec = targetViewportSpec != null
                ? targetViewportSpec
                : ViewportTargetSpec.off();
        this.targetViewportMode = ViewportApplyMode.normalize(targetViewportMode);
        this.targetFontScalePercent = targetFontScalePercent;
        this.targetFontMode = FontApplyMode.normalize(targetFontMode);
        this.targetTypefaceId = targetTypefaceId;
        this.flutterFontHookEnabled = flutterFontHookEnabled;
        this.flutterSettingsFontHookEnabled = flutterSettingsFontHookEnabled;
        this.hyperOsFlutterFontHookEnabled = hyperOsFlutterFontHookEnabled;
        this.hookDomainOverride = hookDomainOverride != null
                ? hookDomainOverride
                : HookDomainOverride.automatic();
    }
}

File: app/src/main/java/com/dpis/module/config/PackageConfigSnapshot.kt
package com.dpis.module.config

import com.dpis.module.fonts.FontApplyMode
import com.dpis.module.hooks.HookDomainOverride
import com.dpis.module.viewport.ViewportApplyMode
import com.dpis.module.viewport.ViewportTargetSpec

class PackageConfigSnapshot(
    packageName: String?,
    @JvmField val dpisEnabled: Boolean,
    targetViewportSpec: ViewportTargetSpec,
    targetViewportMode: String?,
    @JvmField val targetFontScalePercent: Int?,
    targetFontMode: String?,
    @JvmField val targetTypefaceId: String?,
    @JvmField val flutterFontHookEnabled: Boolean,
    @JvmField val flutterSettingsFontHookEnabled: Boolean,
    @JvmField val hyperOsFlutterFontHookEnabled: Boolean,
    hookDomainOverride: HookDomainOverride,
) {
    @JvmField val packageName: String = packageName ?: ""
    @JvmField val targetViewportSpec: ViewportTargetSpec = targetViewportSpec
    @JvmField val targetViewportMode: String = ViewportApplyMode.normalize(targetViewportMode)
    @JvmField val targetFontMode: String = FontApplyMode.normalize(targetFontMode)
    @JvmField val hookDomainOverride: HookDomainOverride = hookDomainOverride

    constructor(
        packageName: String,
        dpisEnabled: Boolean,
        targetViewportWidthDp: Int?,
        targetViewportMode: String?,
        targetFontScalePercent: Int?,
        targetFontMode: String?,
        targetTypefaceId: String?,
        flutterFontHookEnabled: Boolean,
        flutterSettingsFontHookEnabled: Boolean,
        hyperOsFlutterFontHookEnabled: Boolean,
    ) : this(
        packageName,
        dpisEnabled,
        targetViewportWidthDp?.let(ViewportTargetSpec::absoluteDp) ?: ViewportTargetSpec.off(),
        targetViewportMode ?: ViewportApplyMode.OFF,
        targetFontScalePercent,
        targetFontMode ?: FontApplyMode.OFF,
        targetTypefaceId,
        flutterFontHookEnabled,
        flutterSettingsFontHookEnabled,
        hyperOsFlutterFontHookEnabled,
        HookDomainOverride.automatic(),
    )

    constructor(
        packageName: String,
        dpisEnabled: Boolean,
        targetViewportWidthDp: Int?,
        targetViewportMode: String?,
        targetFontScalePercent: Int?,
        targetFontMode: String?,
        targetTypefaceId: String?,
        flutterFontHookEnabled: Boolean,
        flutterSettingsFontHookEnabled: Boolean,
        hyperOsFlutterFontHookEnabled: Boolean,
        hookDomainOverride: HookDomainOverride?,
    ) : this(
        packageName,
        dpisEnabled,
        targetViewportWidthDp?.let(ViewportTargetSpec::absoluteDp) ?: ViewportTargetSpec.off(),
        targetViewportMode,
        targetFontScalePercent,
        targetFontMode,
        targetTypefaceId,
        flutterFontHookEnabled,
        flutterSettingsFontHookEnabled,
        hyperOsFlutterFontHookEnabled,
        hookDomainOverride ?: HookDomainOverride.automatic(),
    )

    constructor(
        packageName: String,
        dpisEnabled: Boolean,
        targetViewportSpec: ViewportTargetSpec,
        targetViewportWidthDp: Int?,
        targetViewportMode: String?,
        targetFontScalePercent: Int?,
        targetFontMode: String?,
        targetTypefaceId: String?,
        flutterFontHookEnabled: Boolean,
        flutterSettingsFontHookEnabled: Boolean,
        hyperOsFlutterFontHookEnabled: Boolean,
        hookDomainOverride: HookDomainOverride,
    ) : this(
        packageName,
        dpisEnabled,
        targetViewportSpec,
        targetViewportMode,
        targetFontScalePercent,
        targetFontMode,
        targetTypefaceId,
        flutterFontHookEnabled,
        flutterSettingsFontHookEnabled,
        hyperOsFlutterFontHookEnabled,
        hookDomainOverride,
    )

}
